File: graphite-server/internal/cypher/ast.go
// Package cypher parses the read-only Cypher language accepted by Graphite.
package cypher

// Query contains UNION branches. UnionAll[i] joins Branches[i] and Branches[i+1].
type Query struct {
	Branches []SingleQuery
	UnionAll []bool
}
type SingleQuery struct{ Clauses []Clause }
type Clause interface{ clause() }
type MatchClause struct {
	Patterns []Pattern
	Optional bool
	Where    Expr
}

func (MatchClause) clause() {}

type ProjectionClause struct {
	With, Distinct, All bool
	Items               []ReturnItem
	Where               Expr
	OrderBy             []SortItem
	Skip, Limit         Expr
}

func (ProjectionClause) clause() {}

type UnwindClause struct {
	Expression Expr
	Variable   string
}

func (UnwindClause) clause() {}

type ReturnItem struct {
	Expression Expr
	Alias      string
	Text       string
}
type SortItem struct {
	Expression Expr
	Descending bool
}
type Pattern struct {
	PathVariable  string
	Nodes         []NodePattern
	Relationships []RelationshipPattern
}
type NodePattern struct {
	Variable   string
	Labels     []string
	Properties map[string]Expr
}
type RelationshipPattern struct {
	Variable         string
	Types            []string
	Properties       map[string]Expr
	Direction        Direction
	VariableLength   bool
	MinHops, MaxHops *int
}
type Direction string

const (
	Outgoing Direction = "outgoing"
	Incoming Direction = "incoming"
	Both     Direction = "both"
)

type Expr interface{ expr() }
type Literal struct{ Value any }

func (Literal) expr() {}

type Variable struct{ Name string }

func (Variable) expr() {}

type Parameter struct{ Name string }

func (Parameter) expr() {}

type Property struct {
	Object Expr
	Key    string
}

func (Property) expr() {}

type Binary struct {
	Op          string
	Left, Right Expr
}

func (Binary) expr() {}

type Unary struct {
	Op      string
	Operand Expr
}

func (Unary) expr() {}

type Call struct {
	Name           string
	Arguments      []Expr
	Distinct, Star bool
}

func (Call) expr() {}

type List struct{ Elements []Expr }

func (List) expr() {}

type Map struct {
	Entries map[string]Expr
	Keys    []string
}

func (Map) expr() {}

type Index struct{ Object, Index Expr }

func (Index) expr() {}

type Slice struct{ Object, From, To Expr }

func (Slice) expr() {}

type Case struct {
	Test  Expr
	Whens []When
	Else  Expr
}

func (Case) expr() {}

type When struct{ Condition, Result Expr }
type ListComprehension struct {
	Variable                string
	List, Where, Projection Expr
}

func (ListComprehension) expr() {}

type Predicate struct {
	Name, Variable string
	List, Where    Expr
}

func (Predicate) expr() {}

// Mutation clauses are parsed for source fidelity. Read-only execution rejects
// these AST nodes; recognizing syntax never authorizes a graph modification.
type CreateClause struct{ Patterns []Pattern }

func (CreateClause) clause() {}

type DeleteClause struct {
	Expressions []Expr
	Detach      bool
}

func (DeleteClause) clause() {}

type SetClause struct{ Items []SetItem }

func (SetClause) clause() {}

type SetItem struct {
	Kind               SetKind
	Variable, Property string
	Expression         Expr
	Labels             []string
}
type SetKind string

const (
	SetProperty        SetKind = "property"
	SetAllProperties   SetKind = "all-properties"
	SetMergeProperties SetKind = "merge-properties"
	SetLabels          SetKind = "labels"
)

type RemoveClause struct{ Items []RemoveItem }

func (RemoveClause) clause() {}

type RemoveItem struct {
	Kind               RemoveKind
	Variable, Property string
	Labels             []string
}
type RemoveKind string

const (
	RemoveProperty RemoveKind = "property"
	RemoveLabels   RemoveKind = "labels"
)

File: backend/cypher/src/ast.rs
//! Cypher AST. Mirrors the Kotlin `CypherClause` / `CypherPattern` / `CypherExpr` model.

#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    Null,
    Bool(bool),
    /// Kotlin `Int` (fits in i32) or `Long`.
    Int(i64),
    Float(f64),
    Str(String),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CmpOp {
    Eq,
    Ne,
    Lt,
    Gt,
    Le,
    Ge,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StrOp {
    StartsWith,
    EndsWith,
    Contains,
    Regex,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Variable(String),
    Parameter(String),
    Property {
        expr: Box<Expr>,
        key: String,
    },
    FunctionCall {
        name: String,
        distinct: bool,
        args: Vec<Expr>,
    },
    CountStar,
    /// `DISTINCT expr` used as an atom.
    Distinct(Box<Expr>),
    Binary {
        op: BinOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Unary {
        negate: bool,
        expr: Box<Expr>,
    },
    Comparison {
        op: CmpOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    StringOp {
        op: StrOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    In {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    IsNull(Box<Expr>),
    IsNotNull(Box<Expr>),
    Not(Box<Expr>),
    And(Box<Expr>, Box<Expr>),
    Or(Box<Expr>, Box<Expr>),
    Xor(Box<Expr>, Box<Expr>),
    Case {
        test: Option<Box<Expr>>,
        whens: Vec<(Expr, Expr)>,
        else_expr: Option<Box<Expr>>,
    },
    ListLiteral(Vec<Expr>),
    MapLiteral(Vec<(String, Expr)>),
    ListComprehension {
        variable: String,
        list: Box<Expr>,
        filter: Option<Box<Expr>>,
        map: Option<Box<Expr>>,
    },
    /// any/all/none/single
    PredicateFunction {
        name: String,
        variable: String,
        list: Box<Expr>,
        predicate: Option<Box<Expr>>,
    },
    Subscript {
        expr: Box<Expr>,
        index: Box<Expr>,
    },
    Slice {
        expr: Box<Expr>,
        from: Option<Box<Expr>>,
        to: Option<Box<Expr>>,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Outgoing,
    Incoming,
    Both,
}

#[derive(Debug, Clone, PartialEq)]
pub struct NodePattern {
    pub variable: Option<String>,
    pub labels: Vec<String>,
    pub properties: Vec<(String, Expr)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RelPattern {
    pub variable: Option<String>,
    pub types: Vec<String>,
    pub direction: Direction,
    pub variable_length: bool,
    pub min_hops: Option<u32>,
    pub max_hops: Option<u32>,
    pub properties: Vec<(String, Expr)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Pattern {
    pub path_variable: Option<String>,
    pub nodes: Vec<NodePattern>,
    /// `rels.len() == nodes.len() - 1`
    pub rels: Vec<RelPattern>,
}

impl Pattern {
    /// All variables introduced by the pattern (path, nodes, relationships) in order.
    pub fn variables(&self) -> Vec<&str> {
        let mut out = Vec::new();
        if let Some(p) = &self.path_variable {
            out.push(p.as_str());
        }
        for (i, n) in self.nodes.iter().enumerate() {
            if let Some(v) = &n.variable {
                out.push(v.as_str());
            }
            if let Some(r) = self.rels.get(i) {
                if let Some(v) = &r.variable {
                    out.push(v.as_str());
                }
            }
        }
        out
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ReturnItem {
    pub expr: Expr,
    pub alias: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OrderItem {
    pub expr: Expr,
    pub descending: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Clause {
    Match {
        patterns: Vec<Pattern>,
        optional: bool,
        /// Only populated for OPTIONAL MATCH; plain MATCH emits a separate Where clause.
        where_clause: Option<Expr>,
    },
    Where(Expr),
    Unwind {
        expr: Expr,
        variable: String,
    },
    With {
        distinct: bool,
        /// None means `WITH *`
        items: Option<Vec<ReturnItem>>,
        where_clause: Option<Expr>,
    },
    Return {
        distinct: bool,
        /// None means `RETURN *`
        items: Option<Vec<ReturnItem>>,
    },
    OrderBy(Vec<OrderItem>),
    Skip(Expr),
    Limit(Expr),
    Create(Vec<Pattern>),
    Delete {
        detach: bool,
        exprs: Vec<Expr>,
    },
    Set,
    Remove,
    /// UNION marker between clause runs; `all` = UNION ALL.
    Union {
        all: bool,
    },
}
